oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/JwtDecoderBuilder.java

```diff
@@ -0,0 +1,223 @@
+import com.nimbusds.jose.jwk.KeyUse;
+import com.nimbusds.jose.jwk.source.JWKSource;
+import com.nimbusds.jose.proc.SecurityContext;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.springframework.security.oauth2.jose.jws.SignatureAlgorithm;
+import org.springframework.util.Assert;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.*;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+/**
+ * An abstraction of the common functionality for the two main JwtDecoderBuilder instances
+ * ({@link NimbusJwtDecoder}, and {@link NimbusReactiveJwtDecoder}).
+ * @param <T> The parent class type
+ *
+ * @author Nick Hitchan
+ */
+public abstract class JwtDecoderBuilder<T> {
+
+	private static final Log log = LogFactory.getLog(JwtDecoderBuilder.class);
+
+	private final String jwkSetUri;
+
+	private final Set<SignatureAlgorithm> signatureAlgorithms = new HashSet<>();
+
+	protected JwtDecoderBuilder(String jwkSetUri) {
+		Assert.hasText(jwkSetUri, "jwkSetUri cannot be empty");
+		this.jwkSetUri = jwkSetUri;
+	}
+
+	protected abstract T self();
+
+	/**
+	 * Provides access to the location of the JWK Set.
+	 * @return the JWK Set URI.
+	 */
+	protected String getJwkSetUri() {
+		return jwkSetUri;
+	}
+
+	/**
+	 * Append the given signing
+	 * <a href="https://tools.ietf.org/html/rfc7515#section-4.1.1" target="_blank">algorithm</a>
+	 * to the set of algorithms to use.
+	 *
+	 * @param signatureAlgorithm the algorithm to use
+	 * @return a {@link NimbusReactiveJwtDecoder.JwkSetUriReactiveJwtDecoderBuilder} for further configurations
+	 */
+	public T jwsAlgorithm(SignatureAlgorithm signatureAlgorithm) {
+		Assert.notNull(signatureAlgorithm, "sig cannot be null");
+		this.signatureAlgorithms.add(signatureAlgorithm);
+		return self();
+	}
+
+	/**
+	 * Configure the list of
+	 * <a href="https://tools.ietf.org/html/rfc7515#section-4.1.1" target="_blank">algorithms</a>
+	 * to use with the given {@link Consumer}.
+	 *
+	 * @param signatureAlgorithmsConsumer a {@link Consumer} for further configuring the algorithm list
+	 * @return a {@link NimbusReactiveJwtDecoder.JwkSetUriReactiveJwtDecoderBuilder} for further configurations
+	 */
+	public T jwsAlgorithms(Consumer<Set<SignatureAlgorithm>> signatureAlgorithmsConsumer) {
+		Assert.notNull(signatureAlgorithmsConsumer, "signatureAlgorithmsConsumer cannot be null");
+		signatureAlgorithmsConsumer.accept(this.signatureAlgorithms);
+		return self();
+	}
+
+	/**
+	 * Fetches {@link SignatureAlgorithm}s based on the configured {@link JWKSource}s keys.
+	 * @param jwkSource
+	 * @return A set of {@link JWSAlgorithm}s to be used for JWT signature verification.
+	 */
+	protected Set<JWSAlgorithm> getSignatureAlgorithms(JWKSource<SecurityContext> jwkSource) {
+		Set<SignatureAlgorithm> jwkAlgorithms = getDefaultAlgorithms();
+		try {
+			jwkAlgorithms.addAll(fetchSignatureVerificationAlgorithms(jwkSource));
```

Comment:
Spring Security builders are not typically additive - instead they replace. This allows Spring Security to gracefully backoff when an application wants to manually configure a value.

What that means here is that if the application has configured any algorithms, then the auto-fetch doesn't get run.

Reply:
Understood, I agree that is definitely a better approach. I will make the required changes.